Llama MLP prefill perf improvement

LPanosTT opened the pull request:
### What's changed
Optimized program configs for all matmuls in the MLP layer for seq_len >= 4k. These changes require an interleaved weight tensor. So, each MLP layer will now have width-shared weights AND interleaved weights with identical data. The interleaved weights will be used during prefill if the sequence length >= 4k.

On 4U, these changes will save ~80ms during prefill at a 4k sequence length.

Perf difference **W1/W3** by sequence length (% reduction in device kernel duration):
- 128: 0%
- 256: 0%
- 512: 0%
- 1024 0%
- 2048 0%
- 4096: 28%
- 6144: 28%
- 8192: 28%
- 10240: 28%
- 12288: 28%
- 14336: 28%
- 16384: 28%
- 24576: 28%
- 32768: 28%
- 51200: 28%
- 65536: 28%
- 86016: 28%
- 131072: 28%

Perf difference **W2** by sequence length (% reduction in device kernel duration):
- 128: ~2%
- 256: 0%
- 512: ~2%
- 1024: 0%
- 2048: 0%
- 4096: 37%
- 6144: 38%
- 8192: 38%
- 10240: 39%
- 12288: 39%
- 14336: 40%
- 16384: 38%
- 24576: 39%
- 32768: 38%
- 51200: 39%
- 65536: 38%
- 86016: 40%
- 131072: 38%

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

LPanosTT committed:
Perform an interleaved matmul in the MLP layer during prefill.
- Store a prefill-only version of the weights (inteleaved) on device in addition to the original which are still used for decode
- Use a 7x7 core grid to prevent hangs
- Only split the seq_len into dim -3 if the seq_len > 2048

This yields a 19% reduction in kernel device duration for 4k - 16k sequence length. The perf is worsened for 128 seq_len, further investigation required to determine why, and at which seq_len this change begins to improve perf.

LPanosTT committed:
special case ff2

Remove commented code

sraizada-tt commented:
Thanks! 2 questions:
1. For the case where you are using the op's default config, can you find out what that config is and pass that explicitly in the matmul so that things don't change unknowingly?
2. For the case where you have DRAM interleaved weights, does passing in the same program config give the same performance? And if it does, we don't need an extra copy of the weights

mtairum commented:
Good stuff!

As @sraizada-tt mentioned, please have either a fixed program_config (or parametrizable) so we know we are running always the same config; otherwise a tt-metal update might change the config and we'll be none the wiser.

Regarding the 2nd point, given that in production we expect any seqlen size from very short to very large (up to 128k) we need to keep both W2 initialized, so I agree with @LPanosTT solution here.

mtairum reviewed (approved):
Pre-approved pending tests and the changes mentioned in comments.

@LPanosTT can you also share in this PR a device trace of 4K seqlen?

LPanosTT committed:
Explicate program configs

LPanosTT commented:
@sraizada-tt 
1. This is dependent on the sequence length and is stored as a lambda in the `model_config`:
    ```
   self.model_config[
                "PREFILL_MLP_W2_PRG_CONFIG"
            ] = lambda seq_len: ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
                compute_with_storage_grid_size=(7, 10),
                in0_block_w=8,  # FIXME: optimize this config for prefill, careful use DI_DT_WORKAROUND if necessary
                out_subblock_h=1,  # Must be divisible by per_core_M
                out_subblock_w=2,  # Must be divisible by per_core_N, out_subblock_w * out_subblock_h <= 4
                per_core_M=max(1, 8 if seq_len >= 2048 else seq_len // self.tile_size // 8),  # 8~10 rows
                per_core_N=math.ceil(2048 / 32 / 7),  # N / TILE_WIDTH / grid width
                transpose_mcast=False,
                fused_activation=None,
                fuse_batch=seq_len <= 2048,
            )
```
2. Using the same program config that we currently use on the interleaved tensors always yields a better result, however in some cases using the one auto generated by the op is even better. I've made changes to explicate which program configs are used.

mtairum commented:
@LPanosTT Can you also comment on why the longer seqlens have a 1D matmul instead of a 2D like the smaller ones?

LPanosTT committed:
hop_cores fix

config fixes

Temporarily disable caching the new weights for CI

LPanosTT committed:
Add 4k seq len mlp prefill test

LPanosTT committed:
experimental change program config generation

Fix bug which reshapes the wrong tensor

do not split the sequence length into the batch dim, this allows us to avoid having to intelligently pick a divisor.

LPanosTT committed:
Change variable name to seq_len since that is what we are using now

LPanosTT committed:
New W1/W3 mm configs

LPanosTT committed:
Update w1/2/3 duration targets

LPanosTT committed:
Fix license headers

LPanosTT committed:
Fix program config gen for W1/W3 so mamtuls fit in L1

LPanosTT committed:
Fix TILE_SIZE to self.tile_size
